tests/unit/test_charm.py

```diff
@@ -1,109 +1,285 @@
-        mock_rm.reconcile.assert_has_calls(
-            [
-                # 3 containers from quantity
-                call("container", 3),
-                call("virtual-machine", 10, VMResources(4, "7GiB", "10GiB")),
-            ]
+        rm.assert_called_with(
+            "github-runner",
+            RunnerManagerConfig(path=GitHubRepo(owner="mockorg", repo="repo"), token="mocktoken"),
+            proxies={},
+        )
+        mock_rm.reconcile.assert_called_with(
+            10, VirtualMachineResources(cpu=4, memory="7GiB", disk="10GiB")
         )
         mock_rm.reset_mock()
 
-    # def test_add_cron(self):
-    #     """Test adding cron jobs."""
-    #     event = "test-event"
-    #     interval = "test-interval"
-    #     harness = Harness(GithubRunnerOperator)
-    #     harness.begin()
-    #     harness.charm._add_cron(event, interval)
-    #     harness.charm._remove_cron(event)
-    #     calls = {}
-    #     for mock_call in harness.charm._cron_tab.mock_calls:
-    #         name, args, kwargs = mock_call
-    #         calls[name] = {"args": args, "kwargs": kwargs}
-
-    #     # Check add cron calls CronTab with expected action and interval
-    #     print(type(calls), "\n")
-    #     print(calls, "\n")
-    #     print(harness.charm._cron_tab, "\n")
-    #     assert calls["().new"]["kwargs"]["command"].split("/")[-1] == event
-    #     assert interval in calls["().new().setall"]["args"]
-    #     # Check that add and remove use same comment
-    #     assert (
-    #         calls["().new"]["kwargs"]["comment"] == \
-    #              calls["().find_comment"]["args"][0]
-    #     )
+    @patch("charm.RunnerManager")
+    @patch("pathlib.Path.write_text")
+    @patch("subprocess.run")
+    def test_on_stop(self, run, wt, rm):
+        rm.return_value = mock_rm = MagicMock()
+        harness = Harness(GithubRunnerCharm)
+        harness.update_config({"path": "mockorg/repo", "token": "mocktoken"})
+        harness.begin()
+        harness.charm.on.stop.emit()
+        mock_rm.flush.assert_called()
+
+    @patch("pathlib.Path.write_text")
+    @patch("subprocess.run")
+    def test_get_runner_manager(self, run, wt):
+        harness = Harness(GithubRunnerCharm)
+        harness.begin()
+
+        # Get runner manager via input.
+        assert harness.charm._get_runner_manager("mocktoken", "mockorg/repo") is not None
+
+        assert harness.charm._get_runner_manager() is None
+
+        # Get runner manager via config.
+        harness.update_config({"path": "mockorg/repo", "token": "mocktoken"})
+        assert harness.charm._get_runner_manager() is not None
+
+        # With invalid path.
+        assert harness.charm._get_runner_manager("mocktoken", "mock/invalid/path") is None
+
+    @patch("charm.RunnerManager")
+    @patch("pathlib.Path.write_text")
+    @patch("subprocess.run")
+    def test_on_install_failure(self, run, wt, rm):
+        """Test various error thrown during install."""
```

Comment:
We have this standard on writing documentation for tests, is it applicable here? https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md#test-structure

Reply:
This file will be rewritten for the migrate to pytest story as charm test are current using `unittest` lib.
I did add some test to make the code coverage pass 85% for now.

Reply:
Yeh maybe we leave it given the tests need to be rewritten, good point